### PR TITLE
[IP-841]: Copy Quote UI feedback

### PR DESCRIPTION
**Pull request by Verony-makesIT**
## Description
The ajax form doesn't post because there is **no response** from the backend.
The problem comes when fetching the 'custom_fields' data.
As far as I understand it, the ( existing?) data is retrieved from the custom_fields db and put into an array.
If there is no data in the db the array cannot be populated and as a result of this there is no "response".
Therefore, in modal_copy_quote, the code
```
if (response.success === 1) {
    window.location = "<?php echo site_url('invoices/view'); ?>/" + response.invoice_id;
}
```
is not executed in order to close the modal window and to display the copied quote view page.
To solve the problem I extracted the snippet (copied the custom_fields code) from mdl_invoice.php and modified it in mdl_quotes.php.

## Related Issue
#841 Copy quote (modal) UI feedback

## Motivation and Context
#841 See description
* Bugfix: Selected quote date is saved
See commits
* Bugfix: Other client can be selected from list
See commits

## Pull Request Checklist

  * [X] My code follows the code formatting guidelines.
  * [X] I have an issue ID for this pull request.
  * [X] I selected the corresponding branch.
  * [X] I have rebased my changes on top of the corresponding branch.

## Issue Type (Please check one or more)

  * [X] Bugfix
  * [ ] Improvement of an existing Feature
  * [ ] New Feature

### CODE REVIEW

**Comment by Verony-makesIT** [comment]
@nielsdrost7: 
I have read your comments in slack ([comment 1](https://invoiceplane.slack.com/archives/C02DQ4VN5/p1677044971968009?thread_ts=1676705070.854219&cid=C02DQ4VN5) and [comment 2](https://invoiceplane.slack.com/archives/C02DQ4VN5/p1677046519771709?thread_ts=1676705070.854219&cid=C02DQ4VN5)) and replied to them in slack ([my answer](https://invoiceplane.slack.com/archives/C02DQ4VN5/p1677087235852319?thread_ts=1676705070.854219&cid=C02DQ4VN5)) as well.
I regret that my proposal (PR) to fix these bugs is now temporarily (?) put "_on hold_" while I think there is _a misunderstanding_.  
The additional bug (see screenshots) occurs in "_Copy Invoice_" **and not in** "_Copy Quote_".
**Commit:**
image

**Bug (in Copy Invoice):**
image

@nielsdrost7, @naui95, @clockwiseq, ...?
**After reading my comments in slack**, can any of the moderators, reviewers clarify to me what else is expected of me?
The most important thing, _in my opinion_, is that issues in IP be fixed asap.....

**Comment by naui95** [comment]
Dear @Verony-makesIT thank you for the further explanations. Your PR AFAIK is not on hold but just pending review. I personally still did not have time to check out the changes and to test them (and eventually get back to you with any question I might have). Please consider that an update has already been planned and at the moment this fix is not included in the upcoming update (see https://github.com/InvoicePlane/InvoicePlane/releases and https://github.com/InvoicePlane/InvoicePlane/milestone/10), which means the review and integration of this PR is - in my opinion - not top priority at the moment. We really appreciate your contributions to the IP community and code base and fully agree with the idea that issues must be fixed.

**Comment by nielsdrost7** [comment]
Same here: I haven't had time to take a look at your PR, @Verony-makesIT 
We appreciate your PR's, but you're going to have to wait some more until we have time to take a look at it

**Comment by naui95** [comment]
closes #841 
closes #1019

**Comment by Verony-makesIT** [comment]
@naui95, 
I just did some testing to see if Copy Quote and Copy Invoice were still working correctly and everything seems to be fine for me. 
So the PR can be considered finished.

File: application/modules/quotes/views/modal_copy_quote.php
Comment (naui95) [inline application/modules/quotes/views/modal_copy_quote.php:66]: I think this should be discussed. Here we are changing the original behavior which was to suggest as creation date also a date in the past. I believe this is over the changes expected with the resolution of the issue. I suggest the change is reverted.

**Review by naui95 — CHANGES_REQUESTED** [review]
Hey, thank you for the PR. The PR works well 👍 there are some changes to be made in my opinion; I've commented the relevant code lines. Some changes go too far and I would revert them to the original solution.

The PR solves more bugs than the original issue, in particular solves:
- feedback on quote copied (needs changes before merging)
- fixes a bug affecting the ability to select an other client when creating a copy of an invoice
- fixes a bug affecting the date chosen by the modal post when copying a quote

For future reference: it would be better to make different issues and PRs for different bugs.

File: application/modules/quotes/models/Mdl_quotes.php
Comment (Verony-makesIT) [reply application/modules/quotes/models/Mdl_quotes.php:220]: @naui95, because I wanted to use the copy quotes function in IP I discovered this problem. Although I am absolutely no programmer I tried to solve this bug by comparing this "**_copy quotes_**" function with the "**_copy invoices_**" function. 
I succeeded by copying the piece of code from the "**_copy invoices_**" function. 
Your proposal certainly seems to me to be a better solution programmatically.
Therefore, isn't it better and desirable to modify your proposal/solution in the "**copy invoices**" function as well?

File: application/modules/quotes/views/modal_copy_quote.php
Comment (Verony-makesIT) [reply application/modules/quotes/views/modal_copy_quote.php:66]: @naui95, I agree that this should be bediscussed. 
I have no problem reverting this change but...
If I find an "old quote/invoice" with the same items as the one I need then I will use this copy functionality provided I only need to change some limited items. 
Anyway, I will usually always change the old quotation/invoice date to the current date. 
In my opinion it makes no sense to keep that old date.
Just my 2 cents.
> I leave it up to you to decide.

File: application/modules/quotes/models/Mdl_quotes.php
Comment (naui95) [reply application/modules/quotes/models/Mdl_quotes.php]: Hi @Verony-makesIT thank you 🙏. Yes, you might be right. I did not look at how the copying with invoices is done. I will check it out this weekend. If the logic should be the same it might be wise to adapt also the copy invoice as you are suggesting. I will let you know.

File: application/modules/quotes/views/modal_copy_quote.php
Comment (naui95) [reply application/modules/quotes/views/modal_copy_quote.php:66]: @Verony-makesIT you actually have a point (or more than one). The behavior you are suggesting is also followed when copying invoices and I suppose is even the expected one.
https://github.com/InvoicePlane/InvoicePlane/blob/bd22ac6f06dd9860d1f48278eaeff6c90674a026/application/modules/invoices/views/modal_copy_invoice.php#L70
My comment is in fact narrow minded, because in principle a PR should only solve the bug/enhancement/etc. to which it is related. This PR solves multiple bugs (thank you!), not all referenced by the issue, and when I went through your code this change did not look as referenced to a bug to me and...therefore the revert suggestion.

Let's see what @nielsdrost7 thinks about merging this change. I believe that it might be wise to keep it in, as @Verony-makesIT pointed out.

File: application/modules/quotes/models/Mdl_quotes.php
Comment (naui95) [inline application/modules/quotes/models/Mdl_quotes.php:226]: @Verony-makesIT I think this part is also necessary and should be included. Am I missing something?

File: application/modules/quotes/models/Mdl_quotes.php
Comment (nielsdrost7) [inline application/modules/quotes/models/Mdl_quotes.php:221]: If it is an empty array [] you're still saving that empty array to the custom_fields for the quote?
Why?

Comment (naui95) [reply application/modules/quotes/models/Mdl_quotes.php:221]: @nielsdrost7 I think we have the same question. I believe that a slice of the original code was accidentally left out (see: https://github.com/InvoicePlane/InvoicePlane/pull/932#discussion_r1527173466)

File: application/modules/quotes/views/modal_copy_quote.php
Comment (nielsdrost7) [reply application/modules/quotes/views/modal_copy_quote.php:66]: @mheiduk Is this another one of those situations where the date suddenly changes? Just wondering.

**Comment by nielsdrost7** [comment]
@Verony-makesIT thank you for your PR.
A couple of general things:
- Normally, don't solve multiple bugs in 1 PR
- Please don't change 'other' things in the same PR as you're solving bugs in.

Right now you're solving bug #932 and feature #1019, right? OK, solved, move on.
Then you're solving something about a client not being set, correct? OK, great, the issue was "copy quote feedback", fine, let's move on.

But wait a minute ... you're also changing a date to today's date after copying the quote?
And that was something that needed to be discussed? 

So now we need to iron out 1 thing that had absolutely nothing to do with solving some bugs? ai, ai.

**Comment by nielsdrost7** [comment]
One last thing, before u have to go over the source itself (indenting problems in the translation file?)

Could you take a look at naui's and my remarks regarding https://github.com/InvoicePlane/InvoicePlane/pull/932#discussion_r1527173466 ?
As soon as some piece of code 'disappears' some bugs may pop up.
I think naui suggested some extra code was needed, focus on that please.
My remark regarding saving the empty array can easily be resolved, but do that after you've solve the other things

File: application/modules/invoices/models/Mdl_invoices.php
Comment (nielsdrost7) [inline application/modules/invoices/models/Mdl_invoices.php:594]: @Verony-makesIT would you mind also removing the extra "enter" on line 619 of this file?

**Comment by nielsdrost7** [comment]
@Verony-makesIT I'll try to make a checklist for you to go over, before we can check the PR one more time and then merge the solution for the Copy Quote UI Feedback:
[] https://github.com/InvoicePlane/InvoicePlane/pull/932/files#r1518791594 : very important. When someone makes a comment and then `requests changes` we cannot merge until those changes have been resolved. I think that single change you made has to be reverted. I can't see which commit it is, but it should be something like `git revert abc13455bca`
[] Fix all the other remarks from naui95, for example here: https://github.com/InvoicePlane/InvoicePlane/pull/932/files#r1527185594 `I believe that a slice of the original code was accidentally left out (see: #932 (comment))`
[] Last but not least: https://github.com/InvoicePlane/InvoicePlane/pull/932/files/7e34967e66f5b62ee34700e34a3a8921271fe3b8#r1528110936 : Remove an "Enter" in a file, since we're changing that file anyway because of your PR, we can remove that "Enter"

Let me know once you're done with these changes, then we can make our final checks.

File: application/modules/invoices/models/Mdl_invoices.php
Comment (Verony-makesIT) [reply application/modules/invoices/models/Mdl_invoices.php:594]: I removed the extra "enter" on line 619 as requested.

File: application/modules/quotes/models/Mdl_quotes.php
Comment (Verony-makesIT) [reply application/modules/quotes/models/Mdl_quotes.php:226]: @naui95 As I've tried to make clear before, I'm absolutely not a programmer and in this case I simply looked at the same function in **Mdl_Invoices.php**, then copied this code into **Mdl_Quotes.php** and tested it. 
Everything worked then without any problems and so I assumed that this is/was (?) the solution for this bug.
Sorry, but I have to admit that I don't have enough (developer) knowledge to answer this question. 
I could be completely wrong, but in my humble opinion this code is not necessary here because it also works in _**Mdl_Invoices.php**_, in the same way and without the "`if (count($db_array) > 2){ }`" code.

File: application/modules/quotes/models/Mdl_quotes.php
Comment (Verony-makesIT) [reply application/modules/quotes/models/Mdl_quotes.php:221]: @naui95 My apologies. 
Here I indeed made a stupid mistake and accidentally modified the code incorrectly. 
In a moment I will restore the code (as I had previously modified it)...

**Review by naui95 — APPROVED** [review]
Fixed copying as proposed and solved conflicts. Ready to go.

**Pull request merged**
